clisp-sources/doc/_clisp.1
.\" ** You probably do not want to edit this file directly **
.\" It was generated using the DocBook XSL Stylesheets (version 2005\-10\-07_10:16_snapshot).
.\" Instead of manually editing it, you probably should edit the DocBook XML
.\" source for it and then use the DocBook XSL Stylesheets to regenerate it.
.TH "CLISP" "1" "Last modified: 2006\-01\-23" "CLISP 2.38" "Platform: @PLATFORM@"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
clisp \- ANSI Common Lisp compiler, interpreter and debugger.
.SH "SYNOPSIS"
.HP 6
\fBclisp\fR [[\fB\-h\fR] [\fB\-\-help\fR]] [\fB\-\-version\fR] [\fB\-\-license\fR] [\fB\-B\ \fIlisp\-lib\-dir\fR\fR]
#ifdef UNIX
[\fB\-K\ \fIlinking\-set\fR\fR]
#endif
[\fB\-M\ \fImem\-file\fR\fR] [\fB\-m\ \fImem\-size\fR\fR]
#ifdef MULTIMAP_MEMORY_VIA_FILE
[\fB\-t\ \fItemp\-dir\fR\fR]
#endif
[\fB\-L\ \fIlanguage\fR\fR] [\fB\-N\ \fIlocale\-dir\fR\fR] [\fB\-E\fIdomain\fR\ \fIencoding\fR\fR] [[\fB\-q\fR] [\fB\-\-quiet\fR] [\fB\-\-silent\fR] [\fB\-v\fR] [\fB\-\-verbose\fR]] [\fB\-on\-error\ \fIaction\fR\fR] [\fB\-repl\fR] [\fB\-w\fR] [\fB\-I\fR] [[\fB\-ansi\fR] [\fB\-traditional\fR]] [\fB\-modern\fR] [\fB\-p\ \fIpackage\fR\fR] [\fB\-C\fR] [\fB\-norc\fR] [\fB\-i\ \fIinit\-file\fR\fR...] [\fB\-c\fR\ [\fB\-l\fR]\ \fIlisp\-file\fR\ [\fB\-o\fR\fB\ \fR\fB\fIoutput\-file\fR\fR]...] [\fB\-x\ \fIexpressions\fR\fR...] [\fB\fIlisp\-file\fR\fR\ [\fB\fIargument\fR\fR...]]
.SH "DESCRIPTION"
.PP
Invokes the
[1]\&\fI\fBCommon Lisp\fR\fR
interpreter and compiler. When called without arguments, executes the
[2]\&\fIread\-eval\-print loop\fR, in which expressions are in turn
[3]\&\fI\fBREAD\fR\fR
from the standard input,
[4]\&\fI\fBEVAL\fR\fRuated by the lisp interpreter, and their results are
[5]\&\fI\fBPRINT\fR\fRed to the standard output. Invoked with
\fB\-c\fR, compiles the specified lisp files to a platform\-independent
bytecode
which can be executed more efficiently.
.SH "OPTIONS"
.PP
\fB\-h\fR, 
.br
\fB\-\-help\fR
.RS
Displays a help message on how to invoke
[6]\&\fI\fBCLISP\fR\fR.
.RE
.TP
\fB\-\-version\fR
Displays the
[6]\&\fI\fBCLISP\fR\fR
version number, as given by the function
[7]\&\fI\fBLISP\-IMPLEMENTATION\-VERSION\fR\fR, the value of the variable
\fI*FEATURES*\fR, as well some other information.
.TP
\fB\-\-license\fR
Displays a summary of the licensing information, the
[8]\&\fIGNU\fR[9]\&\fIGPL\fR.
.TP
\fB\-B\fR \fIlisp\-lib\-dir\fR
Specifies the installation directory. This is the directory containing the linking sets and other data files.
#ifdef UNIX
This option is normally not necessary, because the installation directory is already built\-in into the
\fBclisp\fR
executable.
#endif
Directory
\fIlisp\-lib\-dir\fR
can be changed dynamically using the
[10]\&\fISYMBOL\-MACRO\fR\fI*LIB\-DIRECTORY*\fR.
#ifdef UNIX
.TP
\fB\-K\fR \fIlinking\-set\fR
Specifies the
linking set
to be run. This is a directory (relative to the
\fIlisp\-lib\-dir\fR) containing at least a main executable (runtime) and an initial
memory image. Possible values are
.RS
.TP
\fBbase\fR
the core
[6]\&\fI\fBCLISP\fR\fR
.TP
\fBfull\fR
core plus all the modules with which this installation was built, see
Section\ 31.2, \(lqExternal Modules\(rq.
.RE
.IP
The default is
\fBbase\fR.
#endif
.TP
\fB\-M\fR \fImem\-file\fR
Specifies the initial
memory image. This must be a memory dump produced by the
\fBEXT:SAVEINITMEM\fR
function by this
\fBclisp\fRruntime.
#ifdef UNIX
It may have been compressed using
[8]\&\fIGNU\fR[11]\&\fI\fBgzip\fR\fR.
#endif
.TP
\fB\-m\fR \fImem\-size\fR
Sets the amount of memory
[6]\&\fI\fBCLISP\fR\fR
tries to grab on startup. The amount may be given as
.RS
.TP
\fInnnnnnn\fR
measured in bytes
.PP
\fInnnn\fR\fBK\fR, 
.br
\fInnnn\fR\fBKB\fR
.RS
measured in kilobytes
.RE
.PP
\fIn\fR\fBM\fR, 
.br
\fIn\fR\fBMB\fR
.RS
measured in megabytes
.RE
.RE
.IP
The default is 2 megabytes.
#if (oint_addr_len+addr_shift==24)
The argument is constrained between 100 KB and 16 MB.
#elif (oint_addr_len+addr_shift==26)
The argument is constrained between 100 KB and 64 MB.
#elif (oint_addr_len+addr_shift==28)
The argument is constrained between 100 KB and 256 MB.
#else
The argument is constrained above 100 KB.
#endif
.sp
This version of
[6]\&\fI\fBCLISP\fR\fR
#if defined(SPVW_MIXED) && defined(SPVW_BLOCKS)
#ifdef GENERATIONAL_GC
is not likely to actually use the entire
\fImem\-size\fR
since garbage collection will periodically reduce the amount of used memory. It is therefore common to specify 10 MB even if only 2 MB are going to be used.
#else
eventually uses the entire
\fImem\-size\fR.
#endif
#else
allocates memory dynamically.
\fImem\-size\fR
is essentially ignored (except that it determines the Lisp
STACK
size).
#endif
#ifdef MULTIMAP_MEMORY_VIA_FILE
.TP
\fB\-t\fR \fItemp\-dir\fR
Specifies the directory where the temporary core file is to be held. This option takes precedence over the
[12]\&\fIenvironment variable\fR\fBTMPDIR\fR.
#endif
.TP
\fB\-L\fR \fIlanguage\fR
Specifies the
language[6]\&\fI\fBCLISP\fR\fR
uses to communicate with the user. This may be
#ifndef GNU_GETTEXT
only
\fBenglish\fR.
#else
one of
\fBenglish\fR, \fBgerman\fR, \fBfrench\fR, \fBspanish\fR, \fBdutch\fR, \fBrussian\fR, \fBdanish\fR. Other languages may be specified through the
[12]\&\fIenvironment variable\fR\fBLANG\fR, provided the corresponding message catalog is installed.
#endif
The language may be changed dynamically using the
[10]\&\fISYMBOL\-MACRO\fR\fICUSTOM:*CURRENT\-LANGUAGE*\fR.
.TP
\fB\-N\fR \fIlocale\-dir\fR
Specifies the base directory of locale files.
[6]\&\fI\fBCLISP\fR\fR
will search its message catalogs in
\fI\fIlocale\-dir\fR\fR\fI/\fR\fI\fIlanguage\fR\fR\fI/LC_MESSAGES/clisp.mo\fR. This directory may be changed dynamically using the
[10]\&\fISYMBOL\-MACRO\fR\fICUSTOM:*CURRENT\-LANGUAGE*\fR.
.TP
\fB\-E\fR\fIdomain\fR \fIencoding\fR
Specifies the encoding used for the given domain, overriding the default which depends on the
[12]\&\fIenvironment variable\fRs
\fBLC_ALL\fR,
\fBLC_CTYPE\fR,
\fBLANG\fR.
\fIdomain\fR
can be
.RS
.TP
\fBfile\fR
affecting
\fICUSTOM:*DEFAULT\-FILE\-ENCODING*\fR
.TP
\fBpathname\fR
affecting
\fICUSTOM:*PATHNAME\-ENCODING*\fR
.TP
\fBterminal\fR
affecting
\fICUSTOM:*TERMINAL\-ENCODING*\fR
.TP
\fBforeign\fR
affecting
\fICUSTOM:*FOREIGN\-ENCODING*\fR
.TP
\fBmisc\fR
affecting
\fICUSTOM:*MISC\-ENCODING*\fR
.TP
\fIblank\fR
affecting all of the above.
.RE
.IP

.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBWarning\fR
Note that the values of these
[10]\&\fISYMBOL\-MACRO\fRs that have been saved in a
memory image
are ignored: these
[10]\&\fISYMBOL\-MACRO\fRs are reset based on the OS environment
\fIafter\fR
the
memory image
is loaded. You have to use the
RC file,
\fICUSTOM:*INIT\-HOOKS*\fR
or
init function
to set them on startup, but it is best to set the aforementioned
[12]\&\fIenvironment variable\fRs appropriately for consistency with other programs. See
Section\ 30.1, \(lqCustomizing CLISP Process Initialization and Termination\(rq.
.PP
\fB\-q\fR, 
.br
\fB\-\-quiet\fR, 
.br
\fB\-\-silent\fR, 
.br
\fB\-v\fR, 
.br
\fB\-\-verbose\fR
.RS
Change verbosity level: by default,
[6]\&\fI\fBCLISP\fR\fR
displays a banner at startup and a good\-bye message when quitting, and initializes
[13]\&\fI\fI*LOAD\-VERBOSE*\fR\fR
and
[14]\&\fI\fI*COMPILE\-VERBOSE*\fR\fR
to
[15]\&\fI\fBT\fR\fR, and
[13]\&\fI\fI*LOAD\-PRINT*\fR\fR
and
[14]\&\fI\fI*COMPILE\-PRINT*\fR\fR
to
[16]\&\fI\fBNIL\fR\fR, as per
[ANSI CL standard]. The first
\fB\-q\fR
removes the banner and the good\-bye message, the second sets variables
[13]\&\fI\fI*LOAD\-VERBOSE*\fR\fR
and
[14]\&\fI\fI*COMPILE\-VERBOSE*\fR\fR
to
[16]\&\fI\fBNIL\fR\fR. The first
\fB\-v\fR
sets variables
\fICUSTOM:*REPORT\-ERROR\-PRINT\-BACKTRACE*\fR,
[13]\&\fI\fI*LOAD\-PRINT*\fR\fR
and
[14]\&\fI\fI*COMPILE\-PRINT*\fR\fR
to
[15]\&\fI\fBT\fR\fR, the second sets
\fICUSTOM:*LOAD\-ECHO*\fR
to
[15]\&\fI\fBT\fR\fR. These settings affect the output produced by
\fB\-i\fR
and
\fB\-c\fR
options. Note that these settings persist into the
[2]\&\fIread\-eval\-print loop\fR. Repeated
\fB\-q\fR
and
\fB\-v\fR
cancel each other, e.g.,
\fB\-q \-q \-v \-v \-v\fR
is equivalent to
\fB\-v\fR.
.RE
.TP
\fB\-on\-error\fR \fIaction\fR
Override (or force) the batch mode imposed by
\fB\-c\fR,
\fB\-x\fR, and
\fIlisp\-file\fR, depending on
\fIaction\fR:.RS
.TP
appease
[17]\&\fIcontinuable\fR[18]\&\fIERROR\fRs are turned into
[19]\&\fIWARNING\fRs (with
\fBEXT:APPEASE\-CERRORS\fR) other
[18]\&\fIERROR\fRs are handled in the default way
.TP
debug
[18]\&\fIERROR\fRs
[20]\&\fI\fBINVOKE\-DEBUGGER\fR\fR
(the normal
[2]\&\fIread\-eval\-print loop\fR
behavior)
.TP
abort
[17]\&\fIcontinuable\fR[18]\&\fIERROR\fRs are appeased, other
[18]\&\fIERROR\fRs are
[21]\&\fI\fBABORT\fR\fRed with
\fBEXT:ABORT\-ON\-ERROR\fR
.TP
exit
[17]\&\fIcontinuable\fR[18]\&\fIERROR\fRs are appeased, other
[18]\&\fIERROR\fRs terminate
[6]\&\fI\fBCLISP\fR\fR
with
\fBEXT:EXIT\-ON\-ERROR\fR
.RE
.IP
See also
\fBEXT:SET\-GLOBAL\-HANDLER\fR.
.TP
\fB\-repl\fR
Start an interactive
[2]\&\fIread\-eval\-print loop\fR
after processing the
\fB\-c\fR,
\fB\-x\fR, and
\fIlisp\-file\fR
options and on any
[18]\&\fIERROR\fR[22]\&\fI\fBSIGNAL\fR\fRed during that processing.
.TP
\fB\-w\fR
Wait for a keypress after program termination.
.TP
\fB\-I\fR
Interact better with
[23]\&\fIEmacs\fR
(useful when running
[6]\&\fI\fBCLISP\fR\fR
under
[23]\&\fIEmacs\fR
using
[24]\&\fISLIME\fR,
[25]\&\fIILISP\fR
et al). With this option,
[6]\&\fI\fBCLISP\fR\fR
interacts in a way that
[23]\&\fIEmacs\fR
can deal with:
.RS
.TP 3
\(bu
unnecessary prompts are not suppressed.
.TP
#ifdef GNU_READLINE
\(bu
The
[8]\&\fIGNU\fR[26]\&\fIreadline\fR
library treats
TAB
(see
TAB key) as a normal self\-inserting character (see
Q:\ A.4.3).
#endif
.RE

.TP
\fB\-ansi\fR
Comply with the
[ANSI CL standard]
specification even where
[6]\&\fI\fBCLISP\fR\fR
has been traditionally different. It sets the
[10]\&\fISYMBOL\-MACRO\fR\fICUSTOM:*ANSI*\fR
to
[15]\&\fI\fBT\fR\fR.
.TP
\fB\-traditional\fR
Traditional: reverses the residual effects of
\fB\-ansi\fR
in the saved
memory image.
.TP
\fB\-modern\fR
Provides a modern view of symbols: at startup the
[27]\&\fI\fI*PACKAGE*\fR\fR
variable will be set to the
\(lqCS\-COMMON\-LISP\-USER\(rq
package, and the
[28]\&\fI\fI*PRINT\-BASE*\fR\fR
will be set to
\fB:DOWNCASE\fR. This has the effect that symbol lookup is case\-sensitive (except for keywords and old\-style packages) and that keywords and uninterned symbols are printed with lower\-case preferrence.
.TP
\fB\-p\fR \fIpackage\fR
At startup the value of the variable
[27]\&\fI\fI*PACKAGE*\fR\fR
will be set to the package named
\fIpackage\fR. The default is the value of
[27]\&\fI\fI*PACKAGE*\fR\fR
when the image was
saved, normally
[29]\&\fI\(lqCOMMON\-LISP\-USER\(rq\fR.
.TP
\fB\-C\fR
Compile when loading: at startup the value of the variable
\fICUSTOM:*LOAD\-COMPILING*\fR
will be set to
[15]\&\fI\fBT\fR\fR. Code being
[30]\&\fI\fBLOAD\fR\fRed will then be
[31]\&\fI\fBCOMPILE\fR\fRd on the fly. This results in slower loading, but faster execution.
.TP
\fB\-norc\fR
Normally
[6]\&\fI\fBCLISP\fR\fR
loads the user
[32]\&\fI\(lqrun control\(rq (RC)\fR
file on startup (this happens
\fIafter\fR
the
\fB\-C\fR
option is processed). The file loaded is
\fI.clisprc.lisp\fR
or
\fI.clisprc.fas\fR
in the home directory
[33]\&\fI\fBUSER\-HOMEDIR\-PATHNAME\fR\fR, whichever is newer. This option,
\fB\-norc\fR, prevents loading of the
RC file.
.TP
\fB\-i\fR \fIinit\-file\fR
Specifies initialization files to be
[30]\&\fI\fBLOAD\fR\fRed at startup. These should be lisp files (source or compiled). Several
\fB\-i\fR
options can be given; all the specified files will be loaded in order.
.TP
\fB\-c\fR \fIlisp\-file\fR
Compiles the specified
\fIlisp\-file\fRs to bytecode (\fI*.fas\fR). The compiled files can then be
[30]\&\fI\fBLOAD\fR\fRed instead of the sources to gain efficiency.
.TP
\fB\-o\fR \fIoutputfile\fR
Specifies the output file or directory for the compilation of the last specified
\fIlisp\-file\fR.
.TP
\fB\-l\fR
Produce a bytecode
[34]\&\fI\fBDISASSEMBLE\fR\fR
listing (\fI*.lis\fR) of the files being compiled. Useful only for debugging. See
Section\ 24.1.1, \(lqFunction COMPILE\-FILE\(rq
for details.
.TP
\fB\-x\fR \fIexpressions\fR
Executes a series of arbitrary expressions instead of a
[2]\&\fIread\-eval\-print loop\fR. The values of the expressions will be output to
[35]\&\fI\fI*STANDARD\-OUTPUT*\fR\fR. Due to the argument processing done by the shell, the
\fIexpressions\fR
must be enclosed in double quotes, and double quotes and backslashes must be escaped with backslashes.
.TP
\fIlisp\-file\fR [ \fIargument\fR ... ]
Loads and executes a
\fIlisp\-file\fR, as described in
Script execution. There will be no
[2]\&\fIread\-eval\-print loop\fR. Before
\fIlisp\-file\fR
is loaded, the variable
\fIEXT:*ARGS*\fR
will be bound to a list of strings, representing the
\fIargument\fRs.
#ifdef UNIX
The first line of
\fIlisp\-file\fR
may start with
\fB#!\fR, thus permitting
[6]\&\fI\fBCLISP\fR\fR
to be used as a script interpreter.
#endif
If
\fIlisp\-file\fR
is
\fB\-\fR, the
[35]\&\fI\fI*STANDARD\-INPUT*\fR\fR
is used instead of a file. If
\fIlisp\-file\fR
is the empty string
\fB""\fR
or
\fB"\-\-"\fR, the normal
[2]\&\fIread\-eval\-print loop\fR
is entered, and the rest of the arguments is still available in
\fIEXT:*ARGS*\fR, for parsing by the
init function
of the current
memory image. This option must be the last one. No
RC file
will be executed.
.SH "LANGUAGE REFERENCE"
.PP
The language implemented is
[ANSI CL standard]. The implementation mostly conforms to the
ANSI Common Lisp standard.
.SH "USAGE"
.TP
help
get context\-sensitive on\-line help, see
Chapter\ 25, Environment [CLHS\-25].
.TP

jre/CHANGES
            Java(TM) 2 Runtime Environment, Standard Edition
                             Version 5.0

				CHANGES
------------------------------------------------------------------------

Thank you for downloading this release of the Java(TM) 2 Runtime 
Environment.  The version 5 family of releases contains many new 
features and enhanced functionality compared to previous versions 
of the Java 2 Runtime Environment.  For more information, see:

   * On-line release notes.
     http://java.sun.com/j2se/1.5.0/relnotes.html

   * Summary of New Features - Features added since version 1.4 
     http://java.sun.com/j2se/1.5.0/docs/relnotes/features.html

   * Guide to Features - Complete list of all features.
     http://java.sun.com/j2se/1.5.0/docs/index.html#guide


-----------------------------------------------------------------------
Copyright 2009 Sun Microsystems, Inc., 4150 Network Circle, 
Santa Clara, California 95054, U.S.A. All rights reserved.


clisp-sources/src/_clisp.c
/*
 * Driver program for CLISP.
 *
 * Needed so that we can write #!/usr/local/bin/clisp as the first line
 * of a Lisp script. Normally, only real executables and no shell scripts
 * can be mentioned after #!.
 *
 * Since we are at it, this driver program also implements the "-K" option.
 * All other options are passed to the main program.
 *
 * Bruno Haible 31.3.1997
 * Sam Steingold 1998-2005
 */

/*
 * Macros passed during compilation:
 * LISPLIBDIR  string containing the directory with lisp.run and lispinit.mem
 * LOCALEDIR   string containing the locale directory
 */

#if 0
/*
 * Note: This file is preprocessed. Only the #if's and #include's with
 * no space after the # are resolved. The other ones are preserved by
 * the preprocessing.
 */
#endif

/* Declare strlen(), strcpy(), strcat(). */
# include <string.h>
/* Declare stderr. */
# include <stdio.h>

#if defined(WIN32_NATIVE)
# undef UNICODE
# include <windows.h>
int shell_quote (char * dest, const char * source);
BOOL real_path (LPCSTR namein, LPSTR nameout);
# include "w32shell.c"
#endif

# include <stdlib.h>            /* getenv, abort */
#if defined(UNIX)
/* needed for execname.c to work */
# include <sys/types.h>         /* stat */
# include <sys/stat.h>          /* stat */
# include <unistd.h>            /* access */
# include <errno.h>             /* ENOMEM, ENOENT, errno */
# include <sys/param.h>         /* MAXPATHLEN */
#endif
int find_executable (const char * program_name);
#include "execname.c"

#if !defined(HAVE_PERROR_DECL)
/* Both <errno.h> and <stdio.h> failed to declare perror(). Declare it now. */
# if defined(__cplusplus)
extern "C" void perror (const char *);
# else
extern void perror (const char *);
# endif
#endif

#if defined(UNIX_BINARY_DISTRIB)
# if !ENABLE_RELOCATABLE

char room_for_lisplibdir[10240] = "%MAGIC%LISPLIBDIR=" LISPLIBDIR;
# undef LISPLIBDIR
# define LISPLIBDIR  &room_for_lisplibdir[7 + strlen("LISPLIBDIR") + 1]

char room_for_localedir[10240] = "%MAGIC%LOCALEDIR=" LOCALEDIR;
# undef LOCALEDIR
# define LOCALEDIR  &room_for_localedir[7 + strlen("LOCALEDIR") + 1]

# endif
#endif

int main (int argc, char* argv[])
{
  char* lisplibdir;
  char* localedir;
  char* argv_lisplibdir = NULL;
#if defined(WIN32_NATIVE) && !defined(__MINGW32__)
  char* argv_linkingset = "";
#else
  char* argv_linkingset = "base";
#endif
  char* argv_memfile = NULL;
  char* argv_localedir = NULL;
  char* program_name;
  /*
   * To determine whether -K was given, we go through the options.
   * Because when "clisp foo.lisp -K" is invoked, the "-K" is an argument
   * for foo.lisp, not for clisp. Therefore we have to stop when we encounter
   * the first non-option, non-option-argument. Unfortunately, we have to
   * know about all of clisp's options.
   */
  program_name = argv[0];
  /*
   * The program_name may be absolute or relative if "clisp" is called
   * directly. (For example, "sh /usr/local/bin/clisp ..." will make it
   * absolute, "time clisp ..." will make it relative.)
   * If "clisp" is used as a script interpreter, program_name will be
   * - the full absolute pathname, on SunOS 4, Solaris, HP-UX, IRIX,
   * - only the basename, on Linux, AIX, OSF/1.
   * It follows that we cannot tell whether we have been called as
   * script interpreter or directly.
   */
# if ENABLE_RELOCATABLE
  /* Put this executable's absolute path into executable_name. */
  if (find_executable(program_name) < 0) {
    fprintf(stderr,"%s: cannot figure out the absolute executable path",
            program_name);
    return 1;
  }
  /* Figure out lisplibdir and localedir. */
  {
    unsigned int libdir_len;
    const char *p;
    char *mem;
    /* The libdir is determined as `dirname $executable_name`. */
    for (p = executable_name + strlen(executable_name);; p--) {
      if (p == executable_name) abort();
      if (*p == '/') break;
#ifdef WIN32_NATIVE
      if (*p == '\\') break;
#endif
    }
    libdir_len = p - executable_name;
    mem = (char*)malloc((libdir_len+1)+(libdir_len+7+1));
    if (mem == NULL) goto oom;
    lisplibdir = mem;
    localedir = mem + (libdir_len+1);
    /* Compute lisplibdir from it. */
    memcpy(lisplibdir,executable_name,libdir_len);
    lisplibdir[libdir_len] = '\0';
    /* Compute localedir from it. */
    memcpy(localedir,executable_name,libdir_len);
    localedir[libdir_len] = *p; /* directory separator */
    memcpy(localedir+libdir_len+1,"locale",6);
    localedir[libdir_len+7] = '\0';
  }
# else
  lisplibdir = LISPLIBDIR;
  localedir = LOCALEDIR;
# endif
  /*
   * Script execution on Unix is implemented like this:
   * - The basename/fullname of the interpreter is put into argv[0].
   * - (Optional - only if at least one interpreter-arg is present.) Next
   *   comes the tail of the "#!..." line. On SunOS 4, Linux, IRIX, AIX,
   *   OSF/1: with leading whitespace stripped, but whitespace inside it
   *   untouched (!). On Solaris, HP-UX: with leading whitespace stripped,
   *   and cut off at the next whitespace character (!!).
   * - Next comes the filename of the script.
   * - Then all the arguments of the script.
   * We therefore need to split argv[1] into pieces. We shouldn't split
   * "-x" arguments into pieces. However, fortunately, the "-x" argument
   * cannot appear as argv[1] (because "-x" must precede it), and it
   * cannot appear within the "#!..." line (because "#!" and "-x" are
   * mutually exclusive).
   * As a workaround against the Solaris/HP-UX problem, we split not
   * only at normal spaces, but also at hard spaces.
   * See <impnotes.html#quickstart>.
   */
  if (argc > 1) {
    int wordcount = 0; /* number of pieces in argv[1] */
    { char* arg = argv[1];
      int inword = 0;
      while (*arg != '\0') {
        int spacep = (*arg == '\t' || *arg == ' ' || *arg == (char)0xA0);
        if (!inword && !spacep) wordcount++;
        inword = !spacep;
        arg++;
      }
    }
    {int old_argc = argc;
     char** old_argv = argv;
     int new_argc = argc + wordcount - 1;
     char** new_argv = (char**) malloc((new_argc+1)*sizeof(char*));
     if (!new_argv) goto oom;
     argc = new_argc;
     argv = new_argv;
     /* Copy argv[0] unmodified. */
     *new_argv++ = *old_argv++;
     { /* Split argv[1] into pieces. */
       char* arg = *old_argv++;
       int inword = 0;
       while (*arg != '\0') {
         int spacep = (*arg == '\t' || *arg == ' ' || *arg == (char)0xA0);
         if (!inword) {
           if (!spacep) { *new_argv++ = arg; }
         } else {
           if (spacep) { *arg = '\0'; }
         }
         inword = !spacep;
         arg++;
       }
     }
     { /* Copy argv[2..argc-1] unmodified. */
       int i;
       for (i = old_argc-2; i > 0; i--) { *new_argv++ = *old_argv++; }
     }
     *new_argv = NULL;
    }
  }
  /*
   * Done with script interpreter argument handling.
   */
  { char** argptr = &argv[1];
    char** argptr_limit = &argv[argc];
    enum { for_exec, for_init, for_compile } argv_for = for_exec;
    while (argptr < argptr_limit) {
      char* arg = *argptr++;
      if ((arg[0] == '-') && !(arg[1] == '\0')) {
        switch (arg[1]) {
#        define OPTION_ARG  \
          if (arg[2] == '\0') \
            { if (argptr < argptr_limit) arg = *argptr++; else goto usage; } \
          else { arg = &arg[2]; }
          /* Options to which we have to pay attention. */
          case 'B':
            OPTION_ARG;
            lisplibdir = argv_lisplibdir = arg;
            break;
          case 'K':
            OPTION_ARG;
            argv_linkingset = arg;
            break;
          case 'M':
            OPTION_ARG;
            argv_memfile = arg;
            break;
          case 'N':
            OPTION_ARG;
            argv_localedir = arg;
            break;
          /* Skippable options without arguments. */
          case 'h':
          case 'd':
          case 'q':
          case 'I':
          case 'C':
          case 'l':
          case 'a':
          case 'w':
          case 'n': /* -norc */
          case 'r': /* -repl */
          case 'v':
          case '-':
            break;
          /* Skippable options with arguments. */
          case 'm':
          case 's':
          case 't':
          case 'L':
          case 'o':
          case 'p':
          case 'x':
            OPTION_ARG;
            break;
          case 'E':
            if (argptr < argptr_limit) argptr++; else goto usage;
            break;
          case 'i':
            if (arg[2] == '\0') argv_for = for_init;
            break;
          case 'c':
            argv_for = for_compile;
            break;
          default:
            goto usage;
        }
      } else {
        switch (argv_for) {
          case for_init:
          case for_compile:
            break;
          case for_exec:
            /* All the remaining options are for the Lisp program. */
            goto done_options;
        }
      }
    }
   done_options: ;
  }
  { char* linkingsetdir;
    char* executable;
    char** new_argv;
    /* Compute linking set. */
    if (argv_linkingset == NULL || strlen(argv_linkingset)==0) {
      linkingsetdir = lisplibdir;
    } else if (argv_linkingset[0]=='/') {
      linkingsetdir = argv_linkingset;
    } else {
      linkingsetdir =
        (char*)malloc(strlen(lisplibdir)+1+strlen(argv_linkingset)+1);
      if (!linkingsetdir) goto oom;
      strcpy(linkingsetdir, lisplibdir);
      strcat(linkingsetdir, "/");
      strcat(linkingsetdir, argv_linkingset);
    }
    { /* Compute executable's name. */
#if defined(WIN32_NATIVE) || defined(UNIX_CYGWIN32)
      char* execname = "lisp.exe";
#else
      char* execname = "lisp.run";
#endif
      executable = (char*)malloc(strlen(linkingsetdir)+1+strlen(execname)+1);
      if (!executable) goto oom;
      strcpy(executable, linkingsetdir);
      strcat(executable, "/");
      strcat(executable, execname);
    }
    /* Compute new arguments. */
    new_argv = (char**)malloc((argc+6+1)*sizeof(char*));
    if (!new_argv) goto oom;
    new_argv[0] = executable;
    { char** argptr = &argv[1];
      char** argptr_limit = &argv[argc];
      char** new_argptr = &new_argv[1];
      if (!argv_lisplibdir) {
        *new_argptr++ = "-B";
        *new_argptr++ = lisplibdir;
      }
      if (!argv_memfile) {
        char* filename = "lispinit.mem";
        argv_memfile =
          (char*)malloc(strlen(linkingsetdir)+1+strlen(filename)+1);
        if (!argv_memfile) goto oom;
        strcpy(argv_memfile, linkingsetdir);
        strcat(argv_memfile, "/");
        strcat(argv_memfile, filename);
        *new_argptr++ = "-M";
        *new_argptr++ = argv_memfile;
      }
      if (!argv_localedir) {
        *new_argptr++ = "-N";
        *new_argptr++ = localedir;
      }
      while (argptr < argptr_limit) { *new_argptr++ = *argptr++; }
      *new_argptr = NULL;
    }
    /* Launch the executable. */
#if defined(WIN32_NATIVE)
    {
      PROCESS_INFORMATION pinfo;
      char * command_line = NULL;
      int cmd_len = 0, cmd_pos = 0;
      DWORD exitcode = 0;
      STARTUPINFO sinfo;
      char resolved[MAX_PATH];
      BOOL com_initialized = (CoInitialize(NULL) == S_OK);
      sinfo.cb = sizeof(STARTUPINFO);
      sinfo.lpReserved = NULL;
      sinfo.lpDesktop = NULL;
      sinfo.lpTitle = NULL;
      sinfo.cbReserved2 = 0;
      sinfo.lpReserved2 = NULL;
      sinfo.dwFlags = STARTF_USESTDHANDLES;
      sinfo.hStdInput  = GetStdHandle(STD_INPUT_HANDLE);
      if (sinfo.hStdInput == INVALID_HANDLE_VALUE)  goto w32err;
      sinfo.hStdOutput = GetStdHandle(STD_OUTPUT_HANDLE);
      if (sinfo.hStdOutput == INVALID_HANDLE_VALUE) goto w32err;
      sinfo.hStdError  = GetStdHandle(STD_ERROR_HANDLE);
      if (sinfo.hStdError == INVALID_HANDLE_VALUE)  goto w32err;
      /* conmand line */
      for (argv = new_argv; *argv; argv++) cmd_len += 2*strlen(*argv)+3;
      command_line = (char*)malloc(cmd_len);
      if (!command_line) goto oom;
      command_line[0] = 0;
      for (argv = new_argv; *argv; argv++) {
        if (cmd_pos) command_line[cmd_pos++] = ' ';
        cmd_pos += shell_quote(command_line+cmd_pos,*argv);
        if (cmd_pos > cmd_len) abort();
      }
      if (!CreateProcess((real_path(executable,resolved)
                          ? resolved : executable),
                         command_line, NULL, NULL, 1, 0,
                         NULL, NULL, &sinfo, &pinfo))
        goto w32err;
      /* there is no reason to wait for CLISP to terminate, except that
         if we do not, the console i/o breaks down. */
      if (WaitForSingleObject(pinfo.hProcess,INFINITE) == WAIT_FAILED)
        goto w32err;
      if (!GetExitCodeProcess(pinfo.hProcess,&exitcode)) goto w32err;
      if (!CloseHandle(pinfo.hProcess)) goto w32err;
      if (com_initialized) CoUninitialize();
      return exitcode;
     w32err:
      fprintf(stderr,"%s:\n * %s\n * %s\n * [%s]\n = %s\n",program_name,
              executable,resolved,command_line,strerror(GetLastError()));
      return 1;
    }
#else
    execv(executable,new_argv);
    { /* execv() returns only if there was an error. */
      int saved_errno = errno;
      fprintf(stderr,"%s: ",program_name);
      errno = saved_errno; perror(executable);
    }
#endif
    return 1;
  }
  usage: {
    fprintf(stderr,"%s: invalid command-line option, try `%s --help'\n",
            program_name,program_name);
    return 1;
  }
  oom: {
    fprintf(stderr,"%s: out of memory\n",program_name);
    return 1;
  }
}
